### DISCUSSION TITLE
"Prism EventAggregator" vs "CommunityToolkit Messenger"

### DISCUSSION THREAD

**Discussion by YkTru** [CommunityToolkit/dotnet#780] [Q&A]
1- What are the advantages/disavantages (flexibility, readability, maintainability, functionality etc.) between using Prism EventAggregator/PubSubEvent and CommunityToolkit Messenger?

2- On first sight CommunityToolkit Messenger seems to me easier to implement, and less boilerplate. Is it a more "modern" approach to Prism EventAggregator? Why Messenger has been chosen istead of EventAggregator?

3- I know that CT Messengers has a (great) "Request Message" which Prism EventAggr doesnt have, but other than this are the features equivalent?

Thanks!

**Comment by YkTru** [comment]
(Anyone please? Thank you)

**Reply by Haukinger** [reply]
You have to pay for Prism now, that might count as advantage for CT for many.

**Reply by hailstorm75** [reply]
Source? I was not able to see any clear messaging on Prism's website regarding the necessity for payment.

**Reply by dansiegel** [reply]
> Source? I was not able to see any clear messaging on Prism's website regarding the necessity for payment.


@hailstorm75 it's under the Pricing section of the Prism website. Prism 9.0 is dual license. 

To be clear, you only have to pay for a license if:
- The IP owner earns more than $1 million USD per year
- The IP owner has more than 5 developers which require a license
- The IP owner is a startup (pre-revenue) AND you have received MORE than $3 million USD in funding

_Note that I said IP owner here... this could be an individual or organization. For consultants, their clients are typically the IP owner. This means while the consultant may qualify for the Community License, if their client (who owns the IP) does not they would need a paid license to work on the code for that client._ 

If you do not fall into one of the above exceptions you qualify for the Community License. You do not have to register anywhere, or pay anything and you can continue using Prism for free as you traditionally have. If you are developing an application with an older version of Prism the license can not and does not change from the shipped version meaning that if you are using Prism 6/7/8 it continues to ship under the original license (Apache/MIT depending on the version) and there is nothing you need to do until you update to Prism 9.

The only additional scenario which specifically requires a paid license is if you want access to the additional closed source tooling and plugins and/or access to the private NuGet feed with CI builds to access bug fixes/features as they are merged. In that case you would need the Commercial Plus license.

https://prismlibrary.com/#pricing
https://github.com/PrismLibrary/Prism/blob/master/LICENSE
https://prismlibrary.com/downloads/prism_license.pdf

**Reply by YkTru** [reply]
@hailstorm75 But beware, you'll have to pay top dollar (over $20 a month) for [prism Magician](https://prismmagician.com/) to get modern, decent Prism tooling, even if you're only working on an experimental project, or an **honest** academic project related to medical research, or mathematics. In other words, unless you're an oil company or a bank, or want to make money with your application, you won't get modern, adequate tools. Capitalism at its best, no considerations for poor users, or student. What a shame..

**Reply by dansiegel** [reply]
> you'll have to pay top dollar (over $20 a month)

@YkTru You might want to check the attitude there. First of all, $20 is hardly top dollar. Second for the last several years Prism.Magician (which was developed specifically as Sponsorware), has been available as a thank you to developers who actually have stepped up to make my Open Source work more sustainable with as little as a $10/month GitHub Sponsorship. Btw it's worth considering that GitHub Sponsors doesn't even generate enough cover my rent, and I live in a 3rd world country. At best it covers my flights to speak at maybe two events per year, but doesn't cover my lost income taking the time off to teach developers at conferences or any hotels I may need to stay at. Considering how little I actually get from GitHub Sponsors and the hours that went into developing Prism.Magician it would have over the last several years made my cumulative hourly compensation <$10/hr (even less if you consider the time I had to spend building SponsorConnect)

> Capitalism at its best, no considerations for poor users, or student. What a shame..

If you actually look at the licensing structure you'd realize that it is specifically designed to provide the most we can for free to those who lack financial resources (including students) while requiring large companies who have traditionally benefited from the free use of OSS to actually step up and pay to help solve the problem of OSS Sustainability. I have gone out of my way to make my most of my work freely available to the community and that work has been publicly and freely downloaded in excess of 80 million times on NuGet.org.

**Reply by YkTru** [reply]
@dansiegel  And you think an independent/affiliated researcher earns more than <$10/hr? 

I'm criticizing the fact that you make no difference between a large, lucrative company and a small, independent team, single researcher or student, not the fact that these capitalist (often hearthless) companies should pay you decently for your work. Can't you see the difference between a company making +$300,000-$3M/years, and a student/researcher with scholarship debts from +$30,000$-$100,000? Also, I'm not a US citizen, so your "20$" is worth "30$" in my country..

And I hate the banks, the oil companies, the military, the pharmaceuticals, the marketing companies, and it makes me sick that anyone would decide to encourage or help them *exclusively* at the expense of all the really altruistic, honnest and often important work that many poorly paid researchers do. That's what explains my **legitimate** attitude. Now choose your camp.. (and this doesn't concern Prism.Magician in particular, but your overall "vision" of the Open Source/Community ecosystem (I watched your [video](https://www.youtube.com/watch?v=EU_0QirETEw&pp=ygUWZGFuIHNpZWdlbCBvcGVuIHNvdXJjZQ%3D%3D) and disagree on many points))

**Comment by hailstorm75** [comment]
Using CommunityToolkit in a greenfield project. It is terrific, and sending messages through the application is as easy as 🍰

In general, I would personally not consider any other MVVM framework. CommunityToolkit is fast by design since it utilizes source generators. It provides everything you need for a modern .NET application without any magic.

**Reply by YkTru** [reply]
You haven't tried MvvmGen I guess: https://github.com/thomasclaudiushuber/mvvmgen

I find it more useful in many aspect (though I think I prefer ComTooltkit messengers readability):
Look at the generated/factory viewmodels (for testing or referencing other vm), clean services injection, many different attribute's options for property notification, etc

Anyways I went with Elmish.WPF which is way better than OOP + Prism IMO, not an easy shift but you should check it out!
